system/lib/wasmfs/wasmfs.cpp

```diff
@@ -6,63 +6,94 @@
 // Current Status: Work in Progress.
 // See https://github.com/emscripten-core/emscripten/issues/15041.
 
+#include "file.h"
+#include "file_table.h"
 #include <emscripten/emscripten.h>
 #include <emscripten/html5.h>
+#include <errno.h>
+#include <mutex>
 #include <stdlib.h>
+#include <utility>
 #include <vector>
 #include <wasi/api.h>
 
 extern "C" {
 
-static std::vector<char> fd_write_stdstream_buffer;
+using namespace wasmfs;
+
+__wasi_fd_t __syscall_dup2(__wasi_fd_t oldfd, __wasi_fd_t newfd) {
+  FileTable::Handle fileTable = FileTable::get();
+
+  auto oldOpenFile = fileTable[oldfd];
+  // If oldfd is not a valid file descriptor, then the call fails,
+  // and newfd is not closed.
+  if (!oldOpenFile) {
+    return __WASI_ERRNO_BADF;
+  }
+
+  if (oldfd == newfd) {
+    return oldfd;
+  }
+
+  if (newfd < 0) {
+    return __WASI_ERRNO_BADF;
+  }
+
+  // If the file descriptor newfd was previously open, it is closed
+  // before being reused; the close is performed silently.
+  if (fileTable[newfd]) {
+    fileTable.remove(newfd);
+  }
+
```

Comment:
```suggestion
```
Just overwriting the old entry will have the same effect as explicitly setting it to null before writing to it. This means we can probably get rid of the `remove` method.

Reply:
Hmm ok, I will amend the comment to reflect that as well. I just wanted to originally mimic the POSIX way, but you're right. In this case these is no explicit "closing" since everything is just a `shared_ptr`.